system/lib/wasmfs/js_api.cpp

```diff
@@ -23,6 +23,8 @@ using namespace wasmfs;
 extern "C" {
 
 __wasi_fd_t wasmfs_create_file(char* pathname, mode_t mode, backend_t backend);
+int wasmfs_create_directory(char* path, int mode, backend_t backend);
```

Comment:
Directly using `#include <emscripten/wasmfs.h>` ends up giving me a few weird errors like 
```
C:\Users\james\Documents\GitHub\emscripten\system\lib\wasmfs\js_api.cpp:266:31: error: reference to 'backend_t' is
      ambiguous
  266 | int _wasmfs_mount(char* path, backend_t created_backend) {
      |                               ^
C:\Users\james\Documents\GitHub\emscripten\cache\sysroot/include\emscripten/wasmfs.h:17:25: note: candidate found by
      name lookup is 'backend_t'
   17 | typedef struct Backend* backend_t;
      |                         ^
C:\Users\james\Documents\GitHub\emscripten\system\lib\wasmfs/file.h:35:7: note: candidate found by name lookup is
      'wasmfs::backend_t'
   35 | using backend_t = Backend*;
      |       ^
C:\Users\james\Documents\GitHub\emscripten\system\lib\wasmfs\js_api.cpp:292:10: error: no matching function for call to
      'wasmfs_create_directory'
  292 |   return wasmfs_create_directory(path, 0777, created_backend);
      |          ^~~~~~~~~~~~~~~~~~~~~~~
C:\Users\james\Documents\GitHub\emscripten\cache\sysroot/include\emscripten/wasmfs.h:33:5: note: candidate function not
      viable: no known conversion from 'backend_t' (aka 'wasmfs::Backend *') to 'backend_t' (aka 'Backend *') for 3rd
      argument
   33 | int wasmfs_create_directory(const char* path __attribute__((nonnull)), mode_t mode, backend_t backend);
      |     ^
2 errors generated.
```.

It seems like `backend_t` is redefined in `emscripten/wasmfs.h` using `typedef struct Backend* backend_t` without a namespace, while there is `using namespace wasmfs` in `js_api.cpp`. Would it make sense to change all the usages of `backend_t` in `emscripten/wasmfs.h` to `wasmfs::backend_t`? I'm still new to C++ so this might not be the best approach.

Reply:
Yes, adding the namespace in emscripten/wasmfs.h would fix this. Unfortunately emscripten/wasmfs.h also has to work as a plain C header, though, so it's not entirely straightforward. You would have to insert `namespace wasmfs { ... }` around the typedef only when using C++ and not when using C by using the preprocessor. You can see an example of this in emscripten/proxying.h.

But for this PR, keeping the forward declarations in this file to avoid all these issues seems fine as well.

Reply:
I've tried using the ifdef as shown below
```
#ifdef __cplusplus
namespace wasmfs {
#endif
typedef struct Backend* backend_t;
#ifdef __cplusplus
} // namespace wasmfs
#endif
```

However, it still looks like I'm getting an error with `backend_t` being ambiguous.

```
candidate function not
      viable: no known conversion from 'wasmfs::backend_t' (aka 'wasmfs::Backend *') to 'backend_t' (aka 'Backend *') for
      3rd argument
   35 | int wasmfs_create_directory(const char* path __attribute__((nonnull)), mode_t mode, backend_t backend);
```
Am I using the preprocessor incorrectly? Would it make more sense to use something like `typedef wasmfs::backend_t backend_t` inside the C++ ifdef, then `typedef struct Backend* backend_t` if C++ is not defined? Based on my understanding, doesn't the problem come from having one `backend_t` defined inside the namespace and one that isn't (Instead defined in wasmfs.h)?

Reply:
I was able to solve the problem by wrapping `namespace wasmfs {}` around the entire section, although I'm not sure if this might cause any problems in the future.